lib/faraday/request/retry.rb

```diff
@@ -106,29 +112,35 @@ def initialize(app, options = nil)
       @errmatch = build_exception_matcher(@options.exceptions)
     end
 
-    def sleep_amount(retries)
-      retry_index = @options.max - retries
-      current_interval = @options.interval * (@options.backoff_factor ** retry_index)
-      current_interval = [current_interval, @options.max_interval].min
-      random_interval  = rand * @options.interval_randomness.to_f * @options.interval
-      current_interval + random_interval
+    def calculate_sleep_amount(retries, env)
+      retry_after     = calculate_retry_after(env)
+      retry_interval  = calculate_retry_interval(retries)
+
+      return if retry_after && retry_after > @options.max_interval
+
+      retry_after && retry_after >= retry_interval ? retry_after : retry_interval
     end
 
     def call(env)
       retries = @options.max
       request_body = env[:body]
       begin
         env[:body] = request_body # after failure env[:body] is set to the response body
-        @app.call(env)
+        @app.call(env).tap do |resp|
+          raise Faraday::Error::RetriableResponse.new(nil, resp) if @options.retry_statuses.include?(resp.status)
+        end
       rescue @errmatch => exception
         if retries > 0 && retry_request?(env, exception)
           retries -= 1
           rewind_files(request_body)
           @options.retry_block.call(env, @options, retries, exception)
-          sleep sleep_amount(retries + 1)
-          retry
+          if (sleep_amount = calculate_sleep_amount(retries + 1, env))
+            sleep sleep_amount
+            retry
+          end
         end
-        raise
+
+        raise unless exception.is_a?(Faraday::Error::RetriableResponse)
```

Comment:
@bronislav @iMacTia Hey all, I've tried to use `:retry_statuses`, but it seems that this portion makes it return a nil response and no exceptions are raised when it gives up retrying the `RetriableResponse`, which was unexpected for me. Maybe in that case it would make more sense to return the last `resp`, and then let the user handle the response as he sees fit?

Reply:
@fabiokr I see what you mean. Would you mind to create a pull request for this change?

Reply:
You're right @fabiokr, we missed what to do in case we reach this point (too many retries) with a `Faraday::Error::RetriableResponse`. I agree in that case we should simply return the response, which should be available in the exception.

PR with demonstrating test and fix welcome!

Reply:
@bronislav @iMacTia https://github.com/lostisland/faraday/pull/783